Fixed #28534 -- Make sure changing JSONField in forms always trigger change

shangdahao opened the pull request:
https://code.djangoproject.com/ticket/28534

timgraham commented on django/contrib/postgres/forms/jsonb.py:
```diff
@@ -52,3 +52,12 @@ def prepare_value(self, value):
+        """
+        Return True if data differs from initial.
+        """
+        if super(JSONField, self).has_changed(initial, data):
```
How about:
```
return super(JSONField, self).has_changed(initial, data) or str(initial) != str(data)
```
I think there's no need to repeat the docstring from the superclass but explaining why the `str()` comparison is needed would be useful.

timgraham commented on tests/postgres_tests/test_forms.py:
```diff
@@ -0,0 +1,8 @@
+from . import PostgreSQLTestCase
```
There are existing form field tests in `test_json.py` and `form_fields.py` isn't needed.

shangdahao commented:
Hi @timgraham , I have changed the comparison from str() to json.dumps, because of the order of key in dict may be different, as discuss in the ticket.

And I removed test_forms.py, add tests in test_json.py.

atombrella commented on django/contrib/postgres/forms/jsonb.py:
```diff
@@ -52,3 +52,14 @@ def prepare_value(self, value):
         return json.dumps(value)
+
+    def has_changed(self, initial, data):
+        if super(JSONField, self).has_changed(initial, data):
```
Remove arguments from `super()`.

atombrella commented on django/contrib/postgres/forms/jsonb.py:
```diff
@@ -52,3 +52,14 @@ def prepare_value(self, value):
         return json.dumps(value)
+
+    def has_changed(self, initial, data):
+        if super(JSONField, self).has_changed(initial, data):
```
Remove arguments from `super().`

shangdahao replied:
Done. Thank you!

shangdahao commented on django/contrib/postgres/forms/jsonb.py:
```diff
@@ -52,3 +52,14 @@ def prepare_value(self, value):
+        data_value = data if data is not None else ''
+        # When value in JsonField's item changed from True to 1, or something like that,
+        # for example: {'a': True} -> '{"a": 1}',
+        # they need change to string using json.dumps to comparison.
```
I think the document here is not very good, I will impove it later. Just leave s note here.

shangdahao replied:
I think the document here is not very good, I will impove it later. Just leave s note here

shangdahao commented:
Code and document improved.

orf commented on django/contrib/postgres/forms/jsonb.py:
```diff
@@ -52,3 +52,13 @@ def prepare_value(self, value):
+        # and the order of keys does't matter, so they need to transform to string with same keys order.
+        data_converted = self.to_python(data)
+        initial_value = initial if initial is not None else ''
+        data_value = data_converted if data_converted is not None else ''
```
Is this correct? You're converting `None` to `''`, but wouldn't this miss the case where the value has changed from `null` to `''`, which are distinct values?

shangdahao replied:
Thank you @orf  for review.  The converting here is because I think None and "" in json are the same value, but I am not sure is it correct?
more information(if you need):
The 'initial' parameter has called to_python by its field, so if the field' value is empty, 'initial' is None, if the field's value is "", 'initial' is u''

orf replied:
None in JSON is `null` AFAIK, I think this conditional should be removed if I'm not mistaken?

orf replied:
That being said, if the form value is not set and changes to 'null', this should return true as the value has changed. Maybe add `if data =='null' and initial is None: return True`?

shangdahao replied:
I think you are right,  None and "" in json are not the same value, I will fix this later.

shangdahao commented:
Done. Thank you @orf.

orf commented:
It looks good overall, it still suffers from the issue I added to my comment: if the field is null (so the form field is empty), and the user enters 'null' then this would not return true despite there being a change.

shangdahao commented:
Hi @orf 

The current logic of django is (correct me if I am wrong):

When users enters 'null' or left empty in JSONField of the form, after the form being cleaned, the value will be None, and the None will be saved as NULL in database.

When rendering the change form, if the value of JSONField fetch from database is NULL, it will be converting to None, and The None will be converting to 'null'.

Either the user enter 'null' or left it empty in the form, they are all save as NULL in database, and they are all rendering 'null' in form.

So I think they should treated as not changed.
